### DISCUSSION TITLE
App RouterでextendThemeがエラーを出してしまう

### DISCUSSION THREAD

**Discussion by teram-io** [yamada-ui/yamada-ui#692] [Q&A] [answered]
Next.js 14.1.0、App RouterでYamada UIを使用させていただいております。
themeをカスタマイズしようとしたところ、公式ドキュメント通りのコードでも `extendTheme` を呼び出すとエラーを出してしまうという事象にでくわしています。

エラーを確認しているのは以下のコードです。customThemeを宣言していなければ問題なく動いている状態です。
nextは勉強したてなのでもしコードにミスがあればご教示いただけますと幸いです。

`/app/layout.tsx`
```
import type { Metadata } from "next";
import {
  UIProvider,
  colorModeManager,
  extendTheme,
  ColorModeScript 
} from "@yamada-ui/react";

export const metadata: Metadata = {
  title: "Create Next App",
  description: "Generated by create next app",
};

const customTheme = extendTheme({
  semantics: {
    colors: {
      primary: "pink.500",
    },
    colorSchemes: {
      primary: "pink",
    },
  },
})();


export default function RootLayout({
  children,
}: Readonly<{
  children: React.ReactNode;
}>) {
  return (
   <html lang="en">
      <body>
        <ColorModeScript type="cookie" nonce="testing" />
        <UIProvider
          colorModeManager={{ ...colorModeManager }.cookieStorage}
          theme={customTheme}
        >
          {children}
        </UIProvider>
      </body>
    </html>
  );
}
```

**Comment by illionillion** [comment, marked as answer]
> Next.js 14.1.0、App RouterでYamada UIを使用させていただいております。 themeをカスタマイズしようとしたところ、公式ドキュメント通りのコードでも `extendTheme` を呼び出すとエラーを出してしまうという事象にでくわしています。
> 
> エラーを確認しているのは以下のコードです。customThemeを宣言していなければ問題なく動いている状態です。 nextは勉強したてなのでもしコードにミスがあればご教示いただけますと幸いです。
> 
> `/app/layout.tsx`
> 
> ```
> import type { Metadata } from "next";
> import {
>   UIProvider,
>   colorModeManager,
>   extendTheme,
>   ColorModeScript 
> } from "@yamada-ui/react";
> 
> export const metadata: Metadata = {
>   title: "Create Next App",
>   description: "Generated by create next app",
> };
> 
> const customTheme = extendTheme({
>   semantics: {
>     colors: {
>       primary: "pink.500",
>     },
>     colorSchemes: {
>       primary: "pink",
>     },
>   },
> })();
> 
> 
> export default function RootLayout({
>   children,
> }: Readonly<{
>   children: React.ReactNode;
> }>) {
>   return (
>    <html lang="en">
>       <body>
>         <ColorModeScript type="cookie" nonce="testing" />
>         <UIProvider
>           colorModeManager={{ ...colorModeManager }.cookieStorage}
>           theme={customTheme}
>         >
>           {children}
>         </UIProvider>
>       </body>
>     </html>
>   );
> }
> ```

@Teramon1031 こんにちわ！！

以下のように`extendTheme`の部分を別のファイルに分割して、`use client`をつけた状態でexportしたものを`<UIProvider>`の`theme`に渡してみてください！！

```ts
"use client"
import { extendTheme } from "@yamada-ui/react";

// extendThemeは内部で"use client"で実行しなければならない処理をしているから外部ファイルに"use client"をつけてexportする

export const customTheme = extendTheme({
    semantics: {
        colors: {
            primary: "pink.500",
        },
        colorSchemes: {
            primary: "pink",
        },
    },
})();
```

```tsx
import type { Metadata } from "next";
import { Inter } from "next/font/google";
import { UIProvider } from "@yamada-ui/react"
import { customTheme } from "@/theme";

const inter = Inter({ subsets: ["latin"] });

export const metadata: Metadata = {
  title: "Create Next App",
  description: "Generated by create next app",
};

export default function RootLayout({
  children,
}: Readonly<{
  children: React.ReactNode;
}>) {
  return (
    <html lang="en">
      <body className={inter.className}>
        <UIProvider theme={customTheme}>
          {children}
        </UIProvider>
      </body>
    </html>
  );
}
```

このリポジトリで正常に動作しているのを確かめられるので、git cloneしてみてください。

https://github.com/illionillion/yamada-ui-nextapp-custom-theme
